Skip pods that refer to PVCs that are being deleted

jsafrane opened the pull request:
**What this PR does / why we need it**:

New check was added to `Schedule()` to make sure that a scheduled pod refers to existing PVCs that are not being deleted.

In 1.9 we plan to add a new feature that uses finalizers on PVC to protect PVCs that are used by a running pod from being deleted. This finalizer will be removed when all pods that use a PVC are finished or deleted. See https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md for details.

I needed to pass `pvcLister` to `GenericScheduler`.

UX:

```
$ kubectl describe pod
...
  Type     Reason            Age              From               Message
  ----     ------            ----             ----               -------
  Warning  FailedScheduling  5s (x4 over 8s)  default-scheduler  persistentvolumeclaim "myclaim" is being deleted
  Warning  FailedScheduling  1s (x2 over 1s)  default-scheduler  persistentvolumeclaim "myclaim" not found

```


**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Scheduler skips pods that use a PVC that either does not exist or is being deleted.
```

/sig scheduling
/kind feature

jsafrane commented:
@aveshagarwal, we were talking about this yesteday. Can you please take a look?

aveshagarwal commented:
@jsafrane will take a look soon.

msau42 commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -173,6 +173,13 @@ func defaultPredicates() sets.String {
 			},
 		),
+		// Fit is determined by PVC not being deleted
+		factory.RegisterFitPredicateFactory(
```
The way I added an alpha predicate in my volume scheduling PR was to only register it if the feature gate is set.

msau42 replied:
Nevermind, that method doesn't seem to work.  Ignore me.

liggitt commented:
Will the scheduler schedule a pod that references a PVC that doesn't exist at all?

msau42 commented:
There are other volume predicates that will fail if the PVC doesn't exist.

liggitt commented:
> There are other volume predicates that will fail if the PVC doesn't exist.

Those should treat a PVC with a deletion timestamp as though it doesn't exist, for scheduling purposes

msau42 commented:
As long as this one fails, then the pod will fail to schedule.  It doesn't matter what the other predicates return.  That should be sufficient right?

jsafrane commented:
>> Will the scheduler schedule a pod that references a PVC that doesn't exist at all?

Not with this predicate enabled. It fails with something like `PVC not found` here:

```
		pvc, err := c.pvcInfo.GetPersistentVolumeClaimInfo(namespace, pvcName)
		if err != nil {
			return false, nil, err
```
> There are other volume predicates that will fail if the PVC doesn't exist.

It depends. They work either when unrelated feature is enabled or node has zone/region labels or cloud provider is used. There is no predicate that I could reuse that would work in all clusters, hence I created a new one.

resouer commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1480,3 +1480,49 @@ func (c *VolumeNodeChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetad
+}
+
+// NewPVCPredicate evaluates if a pod can fit due to the PVC it uses.
+func NewNoDeletedPVCsPredicate(pvcInfo PersistentVolumeClaimInfo) algorithm.FitPredicate {
```
Could you please check if this predicate need to be handled seperatedly according to: https://github.com/jsafrane/kubernetes/blob/3835234dd8c78d9955d7e3e06cafaa7197d6a31f/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L71-L77

jsafrane replied:
fixed `invalidatePredicatesForPvc`

jsafrane commented:
rebased & added cache invalidation when PVC changes.

guangxuli commented:
cc/ @guangxuli

jsafrane commented:
/retest

ravisantoshgudimetla commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1480,3 +1480,49 @@ func (c *VolumeNodeChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetad
+	pvcInfo PersistentVolumeClaimInfo
+}
+
+// NewPVCPredicate evaluates if a pod can fit due to the PVC it uses.
```
Need to change the name.

jsafrane replied:
fixed

ravisantoshgudimetla commented:
I agree with you @jsafrane that this shouldn't be a predicate.

/cc @bsalamat for user experience. Is there a way to currently to hold scheduling a pod(like before it starts the calling the schedule f(n)). Changing scheduler name to something else is what I could think of. Anything else that could potentially stop scheduler from trying/re-trying to schedule?

jsafrane commented:
/retest

bsalamat commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1480,3 +1480,49 @@ func (c *VolumeNodeChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetad
+}
+
+func (c *NoDeletedPVCChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PVCProtection) {
```
Why is this feature gated? Isn't it a bug to schedule a pod whose PVC is being deleted?

liggitt replied:
I would have thought so... I was surprised it was allowed today

jsafrane replied:
It changes behavior of scheduler. Currently it can schedule pods that use PVCs that are not bound, their PV counterpart does not exist, they're being deleted or the PVC itself doesn't exist at all.

jsafrane replied:
I removed the gate while moving the check from predicate to a non-node specific check.

bsalamat commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1480,3 +1480,49 @@ func (c *VolumeNodeChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetad
+	return c.predicate
+}
+
+func (c *NoDeletedPVCChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
```
We don't need to run this for every node. Since scheduler doesn't have a better place to put this code, please add a comment with a TODO to mention that it should be run only once per pod.

bsalamat commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1480,3 +1480,49 @@ func (c *VolumeNodeChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMetad
+		}
+
+		if pvc.DeletionTimestamp != nil {
+			return false, nil, fmt.Errorf("PVC %s is being deleted", pvc.Name)
```
If you want to keep this as a predicate, failure reason must be set and error must be nil. The failure reason should be added to the `nodesWherePreemptionMightHelp()` function of generic_scheduler.go as an an unresolvable error.

Another option (recommended one) is to make a function in generic_scheduler.go and call it [here](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/core/generic_scheduler.go#L101). The function can be called something like `podPassesBasicChecks`. If the function returns an error, `Schedule()` function of the generic_scheduler.go returns `"", err` where `err` is the same error as the one returned by `podPassesBasicChecks`. This is a better approach. It won't need changes to other places (such as  `nodesWherePreemptionMightHelp()`) and the check is run only once per pod, not for each node. Future such checks can be added to this newly added function.

msau42 replied:
preempting other pods will not make this pod schedulable.

bsalamat replied:
Yes, that's why this error should be added to `nodesWherePreemptionMightHelp()` as an unresolvable error. Unresolvable errors are those scheduling errors that cannot be addressed by preemption. Well, all of this is needed only if we want to keep this as a predicate.

jsafrane replied:
> make a function in generic_scheduler.go

Done as you suggested. Note that I needed to pass `pvcLister` into `GenericScheduler` and all tests, therefore the PR has grown a bit.

No alpha feature gate!

jsafrane commented:
reworked completely (no predicate) + edited description + release note

jsafrane commented:
and renamed whole PR

jsafrane commented:
rebased
@bsalamat, PTAL if there is still time

bsalamat commented on plugin/pkg/scheduler/algorithm/types.go:
```diff
@@ -117,6 +117,12 @@ type ReplicaSetLister interface {
 }
 
+// PersistentVolumeClaimLister interface represents anything that can list PVCs for a scheduler.
+type PersistentVolumeClaimLister interface {
```
Why do we need this interface? Why can't we pass a PVC Lister to scheduler?

jsafrane replied:
We don't need this interface, but looking around scheduler code, it seems to be a generic pattern to wrap everything in standalone interface instead of using listeners. I am just following the suit.

I'm removing the interface in 2nd commit so we can either squash it or delete it.

bsalamat commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -903,7 +903,7 @@ func (f *configFactory) CreateFromKeys(predicateKeys, priorityKeys sets.String,
 	}
 
-	algo := core.NewGenericScheduler(f.schedulerCache, f.equivalencePodCache, f.podQueue, predicateFuncs, predicateMetaProducer, priorityConfigs, priorityMetaProducer, extenders, f.volumeBinder)
+	algo := core.NewGenericScheduler(f.schedulerCache, f.equivalencePodCache, f.podQueue, predicateFuncs, predicateMetaProducer, priorityConfigs, priorityMetaProducer, extenders, f.volumeBinder, &pvcLister{f.pVCLister})
```
If we didn't have the newly added `PersistentVolumeClaimLister`, couldn't we pass pvcLister directly?

jsafrane replied:
we could, see above

bsalamat reviewed (commented):
We need an integration test. There are many subtle bugs that are hard to catch in unit-tests.
This is a bug fix. So, we should be able to merge it after the code freeze.

jsafrane commented:
/retest

jsafrane committed:
Skip pods that refer to PVCs that are being deleted

Scheduler should ignore pods that refer to PVCs that either do not exist or
are being deleted.

jsafrane committed:
Remove PVCLister and use informer directly.

jsafrane commented on plugin/pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -901,6 +906,36 @@ func podEligibleToPreemptOthers(pod *v1.Pod, nodeNameToInfo map[string]*schedule
+		}
+		pvcName := volume.PersistentVolumeClaim.ClaimName
+		if pvcName == "" {
+			// Theoretically unreachable, this is already ensured by validation.
```
removed (directly in the first commit)

jsafrane commented:
/retest
#56262

bsalamat commented:
/lgtm
Although I preferred the PR to include an integration test.

bsalamat commented:
/approve no-issue
